File: POTD/Day-08_Count-Subarray-with-K-Odds/Solution.java
class Solution {

    // ✅ OPTIMAL APPROACH (Prefix Sum + HashMap)
    public int countSubarrays(int[] arr, int k) {

        HashMap<Integer, Integer> freq = new HashMap<>();
        freq.put(0, 1);   // base case

        int prefixOddCount = 0;
        int count = 0;

        for (int num : arr) {

            if (num % 2 == 1) {
                prefixOddCount++;
            }

            if (freq.containsKey(prefixOddCount - k)) {
                count += freq.get(prefixOddCount - k);
            }

            freq.put(prefixOddCount, freq.getOrDefault(prefixOddCount, 0) + 1);
        }

        return count;
    }
}

/*
================ OTHER APPROACHES (COMMENTED) ================

1️⃣ Sliding Window (At Most K Odds)
- countAtMost(k) - countAtMost(k-1)
- Time: O(n)

2️⃣ Brute Force
- Two loops
- Time: O(n^2)

3️⃣ Full Standalone Runnable Version

class Main {
    public static void main(String[] args) {
        int[] arr = {1, 1, 2, 1, 1};
        int k = 3;
        Solution sol = new Solution();
        System.out.println(sol.countSubarrays(arr, k));
    }
}
*/

File: POTD/Day-08_Count-Subarray-with-K-Odds/Solution.cpp
class Solution {
  public:
    // ✅ OPTIMAL APPROACH (Prefix Sum + Hash Map)
    int countSubarrays(vector<int>& arr, int k) {

        unordered_map<int, int> freq;
        freq[0] = 1;          // base case

        int prefixOddCount = 0;
        int count = 0;

        for (int num : arr) {
            // Treat odd as 1, even as 0
            if (num % 2 == 1) {
                prefixOddCount++;
            }

            // Check if prefixOddCount - k exists
            if (freq.find(prefixOddCount - k) != freq.end()) {
                count += freq[prefixOddCount - k];
            }

            freq[prefixOddCount]++;
        }

        return count;
    }
};

/*
================ OTHER APPROACHES (COMMENTED) ================

1️⃣ Sliding Window (At Most K Odds Trick)
- Exactly K = AtMost(K) - AtMost(K-1)
- Time: O(n)
- Space: O(1)

2️⃣ Brute Force
- Count odds for every subarray
- Time: O(n^2)
- Space: O(1)

3️⃣ Full Standalone Runnable Version

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

int main() {
    vector<int> arr = {1, 1, 2, 1, 1};
    int k = 3;
    Solution sol;
    cout << sol.countSubarrays(arr, k);
    return 0;
}
*/
